src/App.jsx

```diff
@@ -1,248 +1,278 @@
-            >
-              User 1
-            </a>
-
-            <a
-              data-cy="FilterUser"
-              href="#/"
-              className="is-active"
-            >
-              User 2
-            </a>
-
-            <a
-              data-cy="FilterUser"
-              href="#/"
-            >
-              User 3
-            </a>
-          </p>
-
-          <div className="panel-block">
-            <p className="control has-icons-left has-icons-right">
-              <input
-                data-cy="SearchField"
-                type="text"
-                className="input"
-                placeholder="Search"
-                value="qwe"
-              />
-
-              <span className="icon is-left">
-                <i className="fas fa-search" aria-hidden="true" />
-              </span>
-
-              <span className="icon is-right">
-                {/* eslint-disable-next-line jsx-a11y/control-has-associated-label */}
-                <button
-                  data-cy="ClearButton"
-                  type="button"
-                  className="delete"
-                />
-              </span>
+import usersFromServer from './api/users';
+import categoriesFromServer from './api/categories';
+import productsFromServer from './api/products';
+
+const products = productsFromServer.map(product => ({
+  ...product,
+  category: categoriesFromServer
+    .find(category => category.id === product.categoryId),
+  user: usersFromServer
+    .find(user => categoriesFromServer
+      .find(category => category.id === product.categoryId).ownerId
+        === user.id),
+}));
+
+function getPreparedProducts(items, { user, searchedProduct, category }) {
+  let preparedProducts = [...items];
+
+  if (user) {
+    preparedProducts = preparedProducts
+      .filter(product => product.user.id === user);
+  }
+
+  if (searchedProduct) {
+    const normalizedSearchedProduct = searchedProduct.trim().toLowerCase();
+
+    preparedProducts = preparedProducts
+      .filter(product => product.name
+        .toLowerCase().includes(normalizedSearchedProduct));
+  }
+
+  if (category.length !== 0) {
+    preparedProducts = preparedProducts
+      .filter(product => category.includes(product.categoryId));
+  }
+
+  return preparedProducts;
+}
+
+export const App = () => {
+  const [sortProducts, setSortProducts] = useState({
+    user: null,
+    searchedProduct: null,
+    category: [],
+  });
+  const visibleProducts = getPreparedProducts(products, sortProducts);
+
+  const updateSortProductsKey = (key, newValue) => {
+    const updateProducts = { ...sortProducts };
+
+    if (key === 'category' && !(newValue instanceof Array)) {
+      updateProducts[key] = updateProducts[key].includes(newValue)
+        ? updateProducts[key].filter(value => value !== newValue)
+        : [...updateProducts[key], newValue];
+    } else {
+      updateProducts[key] = newValue;
+    }
+
+    setSortProducts(updateProducts);
+  };
```

Comment:
```suggestion
  const updateSortProductsKey = (key, newValue) => {
    const updateProducts = { ...sortProducts };

    if (key === 'category' && !(newValue instanceof Array)) {
      updateProducts[key] = updateProducts[key].includes(newValue)
        ? updateProducts[key].filter(value => value !== newValue)
        : [...updateProducts[key], newValue];
        return;
    }
    
     updateProducts[key] = newValue;

    setSortProducts(updateProducts);
  };
```

Reply:
The solution you proposed doesn't work ☹